### PR TITLE
feat(datastore): Adding x-goog-request-params header

### PR DESCRIPTION
**Pull request by bhshkh**
When using named databases, dynamic routing header needs to be provided in the format `project_id=$project id&database_id=$databaseid`

### CODE REVIEW

**Review by enocom — COMMENTED** [review]
Should we add an integration test for this to avoid regressions?

**Comment by kolea2** [comment]
> Should we add an integration test for this to avoid regressions?

+1, adding a test here would be great!

**Comment by bhshkh** [comment]
> Should we add an integration test for this to avoid regressions?

Added

**Review by enocom — APPROVED** [review]

**Comment by bhshkh** [comment]
@viacheslav-rostovtsev can you please review this?

File: datastore/client.go [hunk 1]
```diff
@@ -47,10 +48,13 @@ func newDatastoreClient(conn grpc.ClientConnInterface, projectID, databaseID str
 	if databaseID != "" {
 		resourcePrefixValue += "/databases/" + databaseID
 	}
+	reqParamsHeaderValue := fmt.Sprintf("project_id=%s&database_id=%s", url.QueryEscape(projectID), url.QueryEscape(databaseID))
```
[the remark this answers is not in the document]
Comment (bhshkh) [reply datastore/client.go:51]: If database id is removed, there are multiple integration test failures:
```
.....
.....
=== RUN   TestIntegration_EventualConsistency
    headers_enforcer.go:184: For method "/google.datastore.v1.Datastore/Commit", errors:
        header "x-goog-request-params": got project_id=cndb-sdk-golang-general, want project_id=cndb-sdk-golang-general&database_id=
--- FAIL: TestIntegration_EventualConsistency (0.00s)
=== RUN   TestIntegration_Projection
    headers_enforcer.go:184: For method "/google.datastore.v1.Datastore/Commit", errors:
        header "x-goog-request-params": got project_id=cndb-sdk-golang-general, want project_id=cndb-sdk-golang-general&database_id=
--- FAIL: TestIntegration_Projection (0.00s)
=== RUN   TestIntegration_AllocateIDs
    headers_enforcer.go:184: For method "/google.datastore.v1.Datastore/AllocateIds", errors:
        header "x-goog-request-params": got project_id=cndb-sdk-golang-general, want project_id=cndb-sdk-golang-general&database_id=
--- FAIL: TestIntegration_AllocateIDs (0.00s)

....
....

```

[the remark this answers is not in the document]
Comment (bhshkh) [reply datastore/client.go:51]: Just to clarify, by remove database_id,
I meant that even if the default database is being used, it needs to be added in the header.

[the remark this answers is not in the document]
Comment (bhshkh) [reply datastore/client.go]: Updated the code as per the spec. Good catch!

File: datastore/client.go [hunk 2]
```diff
@@ -44,13 +45,18 @@ type datastoreClient struct {
 
 func newDatastoreClient(conn grpc.ClientConnInterface, projectID, databaseID string) pb.DatastoreClient {
 	resourcePrefixValue := "projects/" + projectID
-	if databaseID != "" {
+	reqParamsHeaderValue := "project_id=" + url.QueryEscape(projectID)
+
+	if databaseID != DefaultDatabaseID {
```
Comment (kolea2) [inline datastore/client.go:50]: I think you can also keep the empty check

File: datastore/client.go [hunk 3]
```diff
@@ -44,13 +45,18 @@ type datastoreClient struct {
...
 		resourcePrefixValue += "/databases/" + databaseID
+		reqParamsHeaderValue += "&database_id=" + databaseID
```
Comment (kolea2) [inline datastore/client.go:52]: should be wrapped in `url.QueryEscape`?

File: datastore/client.go (hunk 2 above)
Comment (bhshkh) [reply datastore/client.go]: Done

File: datastore/client.go (hunk 3 above)
Comment (bhshkh) [reply datastore/client.go:52]: Done

Comment (kolea2) [reply datastore/client.go:52]: sorry not sure if I missed something, shouldn't `reqParamsHeaderValue` be escaped?

File: datastore/client.go [hunk 4]
```diff
@@ -43,14 +44,19 @@ type datastoreClient struct {
...
-	if databaseID != "" {
-		resourcePrefixValue += "/databases/" + databaseID
+	resourcePrefixValue := "projects/" + url.QueryEscape(projectID)
+	reqParamsHeaderValue := "project_id=" + url.QueryEscape(projectID)
+
+	if databaseID != DefaultDatabaseID && databaseID != "" {
+		resourcePrefixValue += "/databases/" + url.QueryEscape(databaseID)
+		reqParamsHeaderValue += "&database_id=" + url.QueryEscape(databaseID)
 	}
+	fmt.Printf("resourcePrefixValue: %v, reqParamsHeaderValue: %v\n", resourcePrefixValue, reqParamsHeaderValue)
```
Comment (kolea2) [inline datastore/client.go:54]: prob want to remove?

File: datastore/client.go (hunk 3 above)
Comment (bhshkh) [reply datastore/client.go]: Done

File: datastore/client.go (hunk 4 above)
Comment (bhshkh) [reply datastore/client.go]: Sorry about this. Removed.

**Review by kolea2 — APPROVED** [review]

**Pull request merged**